control-plane/gateways/service.go

```diff
@@ -36,15 +35,39 @@ func (b *meshGatewayBuilder) Service() *corev1.Service {
 		Spec: corev1.ServiceSpec{
 			Selector: b.labelsForDeployment(),
 			Type:     serviceType,
-			Ports: []corev1.ServicePort{
-				{
-					Name: "wan",
-					Port: port,
-					TargetPort: intstr.IntOrString{
-						IntVal: port + portModifier,
-					},
+			Ports:    b.Ports(portModifier),
+		},
+	}
+}
+
+// Ports build a list of ports from the listener objects. In theory there should only ever be a WAN port on
+// mesh gateway but building the ports from a list of listeners will
+func (b *meshGatewayBuilder) Ports(portModifier int32) []corev1.ServicePort {
+
+	ports := []corev1.ServicePort{}
+
+	if len(b.gateway.Spec.Listeners) == 0 {
+		//If empty use the default value. This should always be set, but in case it's not, this check
+		//will prevent a panic.
```

Comment:
Can we instead add a minimum length requirement to the listeners field on our CRD using kubebuilder annotations? This would make our code simpler

Reply:
But wouldn't that make it harder to test?

Reply:
I personally would argue that we should be doing both, since the struct where I will add the kubetag is in a different repository but I will defer to others opinions here.

Reply:
https://github.com/hashicorp/consul/pull/20120

PR to take care of the kubetag